src/components/Tooltip/TooltipRenderedOnPageBody.js

```diff
@@ -39,59 +34,99 @@ const propTypes = {
-        tooltipTextStyle,
-        pointerWrapperStyle,
-        pointerStyle,
-    } = getTooltipStyles(
-        props.animation,
-        props.windowWidth,
-        props.xOffset,
-        props.yOffset,
-        props.wrapperWidth,
-        props.wrapperHeight,
-        props.tooltipWidth,
-        props.tooltipHeight,
-        props.shiftHorizontal,
-        props.shiftVertical,
-    );
-    return ReactDOM.createPortal(
-        <Animated.View
-            ref={props.setTooltipRef}
-            onLayout={props.measureTooltip}
-            style={[tooltipWrapperStyle, animationStyle]}
-        >
-            <Text style={tooltipTextStyle} numberOfLines={1}>{props.text}</Text>
-            <View style={pointerWrapperStyle}>
-                <View style={pointerStyle} />
-            </View>
-        </Animated.View>,
-        document.querySelector('body'),
-    );
+    /** Maximum number of lines to set on tooltip */
+    numberOfLines: PropTypes.number.isRequired,
 };
 
-TooltipRenderedOnPageBody.propTypes = propTypes;
-TooltipRenderedOnPageBody.defaultProps = defaultProps;
-TooltipRenderedOnPageBody.displayName = 'TooltipRenderedOnPageBody';
-
 // Props will change frequently.
 // On every tooltip hover, we update the position in state which will result in re-rendering.
 // We also update the state on layout changes which will be triggered often.
 // There will be n number of tooltip components in the page.
 // Its good to memorize this one.
-export default memo(TooltipRenderedOnPageBody);
+class TooltipRenderedOnPageBody extends React.PureComponent {
+    constructor(props) {
+        super(props);
+        this.state = {
+            // The width of toolttip's inner text
+            tooltipTextWidth: 0,
+
+            // The width and height of the tooltip itself
+            tooltipWidth: 0,
+            tooltipHeight: 0,
+        };
+
+        this.measureTooltip = this.measureTooltip.bind(this);
+    }
+
+    componentDidMount() {
+        this.setState({
+            tooltipTextWidth: this.textRef.offsetWidth,
+        });
+    }
+
+    /**
+     * Measure the size of the tooltip itself.
+     *
+     * @param {Object} nativeEvent
+     */
+    measureTooltip({nativeEvent}) {
+        this.setState({
+            tooltipWidth: nativeEvent.layout.width,
+            tooltipHeight: nativeEvent.layout.height,
+        });
+    }
+
+    render() {
+        // We get wrapper width based on tooltip's inner text width so the wrapper
+        // is just big enough to fit text and prevent white space
+        const wrapperWidth = this.state.tooltipTextWidth && this.state.tooltipTextWidth < this.props.maxWidth
+
+            // Add horizontal padding to the text width to get the wrapper width.
+            // TooltipTextWidth ignores the fractions (OffsetWidth) so add 1px to fit the text properly.
+            ? this.state.tooltipTextWidth + (spacing.ph2.paddingHorizontal * 2) + 1
+            : this.props.maxWidth;
```

Comment:
First of all, move this to `getTooltipStyles`.

Reply:
Done, I also changed the params order because I noticed the last two params (`manualShiftHorizontal `and `manualShiftVertical`) are optional.